### ISSUE TITLE
DeFiChain Jellyfish RPC Implementations

### ISSUE THREAD

**Issue by fuxingloh** [DeFiCh/jellyfish#48]
> This tracks all the RPC methods and categories being implemented in `jellyfish-core`. Will list all the methods here, and strike out those that are not applicable.

### `blockchain.cpp`
- [ ] getblockchaininfo
- [ ] getchaintxstats
- [ ] getblockstats
- [ ] getbestblockhash
- [ ] getblockcount
- [ ] getblock
- [ ] getblockhash
- [ ] getblockheader
- [ ] getchaintips
- [ ] getdifficulty
- [ ] getmempoolancestors
- [ ] getmempooldescendants
- [ ] getmempoolentry
- [ ] getmempoolinfo
- [ ] clearmempool
- [ ] getrawmempool
- [ ] gettxout
- [ ] gettxoutsetinfo
- [ ] pruneblockchain
- [ ] savemempool
- [ ] verifychain
- [ ] preciousblock
- [ ] scantxoutset
- [ ] getblockfilter
- [ ] invalidateblock
- [ ] reconsiderblock
- [ ] waitfornewblock
- [ ] waitforblock
- [ ] waitforblockheight
- [ ] syncwithvalidationinterfacequeue

### `mining.ccp`
* [x] getnetworkhashps
* [x] getmintinginfo
* [ ] prioritisetransaction
* [ ] getblocktemplate
* [ ] submitblock
* [ ] submitheader
* [ ] generatetoaddress
* [ ] estimatesmartfee
* [ ] estimaterawfee

### `misc.cpp`
- [ ] getmemoryinfo
- [ ] logging
- [ ] validateaddress
- [ ] createmultisig
- [ ] deriveaddresses
- [ ] getdescriptorinfo
- [ ] verifymessage
- [ ] signmessagewithprivkey
- [ ] setmocktime
- [ ] echo
- [ ] echojson

### `net.cpp`
- [ ] getconnectioncount
- [ ] ping
- [ ] getpeerinfo
- [ ] addnode
- [ ] disconnectnode
- [ ] getaddednodeinfo
- [ ] getnettotals
- [ ] getnetworkinfo
- [ ] setban
- [ ] listbanned
- [ ] clearbanned
- [ ] setnetworkactive
- [ ] getnodeaddresses

### `rawtransaction.cpp`
- [ ] getrawtransaction
- [ ] createrawtransaction
- [ ] decoderawtransaction
- [ ] decodescript
- [ ] sendrawtransaction
- [ ] combinerawtransaction
- [ ] signrawtransactionwithkey
- [ ] testmempoolaccept
- [ ] decodepsbt
- [ ] combinepsbt
- [ ] finalizepsbt
- [ ] createpsbt
- [ ] converttopsbt
- [ ] utxoupdatepsbt
- [ ] analyzepsbt
- [ ] joinpsbts
- [ ] gettxoutproof
- [ ] verifytxoutproof

### `server.cpp`
- [ ] getrpcinfo
- [ ] help
- [ ] stop
- [ ] uptime

### `spv_rpc.cpp`
- [ ] spv_sendrawtx
- [ ] spv_createanchor
- [ ] spv_createanchortemplate
- [ ] spv_estimateanchorcost
- [ ] spv_rescan
- [ ] spv_syncstatus
- [ ] spv_gettxconfirmations
- [ ] spv_splitutxo
- [ ] spv_listanchors
- [ ] spv_listanchorauths
- [ ] spv_listanchorrewardconfirms
- [ ] spv_listanchorrewards
- [ ] spv_listanchorsunrewarded
- [ ] spv_listanchorspending
- [ ] spv_setlastheight

### `rpcwallet.cpp`
- [ ] fundrawtransaction
- [ ] abandontransaction
- [ ] abortrescan
- [ ] addmultisigaddress
- [ ] backupwallet
- [ ] bumpfee
- [ ] createwallet
- [ ] dumpprivkey
- [ ] dumpwallet
- [ ] encryptwallet
- [ ] getaddressesbylabel
- [ ] getaddressinfo
- [x] getbalance
- [ ] getnewaddress
- [ ] getrawchangeaddress
- [ ] getreceivedbyaddress
- [ ] getreceivedbylabel
- [ ] gettransaction
- [ ] getunconfirmedbalance
- [ ] getbalances
- [ ] getwalletinfo
- [ ] importaddress
- [ ] importmulti
- [ ] importprivkey
- [ ] importprunedfunds
- [ ] importpubkey
- [ ] importwallet
- [ ] keypoolrefill
- [ ] listaddressgroupings
- [ ] listlabels
- [ ] listlockunspent
- [ ] listreceivedbyaddress
- [ ] listreceivedbylabel
- [ ] listsinceblock
- [ ] listtransactions
- [ ] listunspent
- [ ] listwalletdir
- [ ] listwallets
- [ ] loadwallet
- [ ] lockunspent
- [ ] removeprunedfunds
- [ ] rescanblockchain
- [ ] sendmany
- [ ] sendtoaddress
- [ ] sethdseed
- [ ] setlabel
- [ ] settxfee
- [ ] setwalletflag
- [ ] signmessage
- [ ] signrawtransactionwithwallet
- [ ] unloadwallet
- [ ] walletcreatefundedpsbt
- [ ] walletlock
- [ ] walletpassphrase
- [ ] walletpassphrasechange
- [ ] walletprocesspsbt

### `mn_rpc.cpp`
- [ ] setgov
- [ ] getgov
- [ ] isappliedcustomtx

### `rpc_accounts.cpp`
- [ ] listaccounts
- [ ] getaccount
- [ ] gettokenbalances
- [ ] utxostoaccount
- [ ] accounttoaccount
- [ ] accounttoutxos
- [ ] listaccounthistory
- [ ] accounthistorycount
- [ ] listcommunitybalances
- [ ] sendtokenstoaddress

### `rpc_masternodes.cpp`
- [ ] createmasternode
- [ ] resignmasternode
- [ ] listmasternodes
- [ ] getmasternode
- [ ] getmasternodeblocks
- [ ] listcriminalproofs
- [ ] getanchorteams
- [ ] getactivemasternodecount
- [ ] listanchors

### `rpc_poolpair.cpp`
- [ ] listpoolpairs
- [ ] getpoolpair
- [ ] addpoolliquidity
- [ ] removepoolliquidity
- [ ] createpoolpair
- [ ] updatepoolpair
- [ ] poolswap
- [ ] listpoolshares
- [ ] testpoolswap

### `rpc_tokens.cpp`
- [ ] createtoken
- [ ] updatetoken
- [ ] listtokens
- [ ] gettoken
- [ ] getcustomtx
- [ ] minttokens

**Comment by fuxingloh** [comment]
Working on `blockchain.cpp`

1. getblockchaininfo

**Comment by canonbrother** [comment]
Working on `blockchain.app`
1. getblock

**Comment by canonbrother** [comment]
Working on `blockhain.cpp`
1. gettxxout

**Comment by fuxingloh** [comment]
Will be working on various `rawtransaction.cpp` rpc for wallet integration.

1. getrawtransaction
1. createrawtransaction
1. decoderawtransaction
1. decodescript
1. sendrawtransaction
1. combinerawtransaction
1. signrawtransactionwithkey

**Comment by fuxingloh** [comment]
Updated RPC typing prioritization, removed prioritization for RPC that are going to be created with `jellyfish-transaction`.

All RPC without side effect/mutating properties (`get`, `list`) should be prioritized over those with side effect (`create`, `update`)

**Comment by fuxingloh** [comment]
To prioritize:

`rpc_tokens.cpp`
- listtokens
- gettoken

`rpc_poolpair.cpp`
- listpoolpairs
- getpoolpair
- listpoolshares

`rpc_accounts.cpp`
- listaccounts
- getaccount
- gettokenbalances
- listaccounthistory

`mining.ccp`
- estimatesmartfee
- estimaterawfee

**Comment by canonbrother** [comment]
Working on (late update)
`blockhain.cpp` 
1. getrawmempool

`wallet.cpp`
2. wallet.getNewAddress
3. wallet.getAddressInfo
4. wallet.validateAddress
5. wallet.listAddressGroupings
6. wallet.getWalletInfo
7. wallet.setWalletFlag
8. wallet.sendToAddress
9. wallet.createWallet

**Comment by canonbrother** [comment]
Working `rpc_tokens.cpp`
1. listokens
2. gettoken

**Comment by canonbrother** [comment]
Working on rpc_poolpair.cpp
1. listpoolpairs
2. getpoolpair
3. listpoolshares

**Comment by canonbrother** [comment]
Working on `rpc_poolpair.cpp` (on hold first, RPC is added but test as no data yet)
1. listPoolPairs
2. getPoolPair
3. listPoolShares

**Comment by canonbrother** [comment]
Working `rpc_accounts.cpp`
1. listaccounts
2. getaccount
3. gettokenbalances
4. listaccounthistory

**Comment by jingyi2811** [comment]
Working on net.cpp

getconnectioncount
getpeerinfo
getnetworkinfo

**Comment by fuxingloh** [comment]
@jingyi2811 mentioned that it's hard to test multi node setup (e.g. `getpeerinfo`) without https://github.com/DeFiCh/jellyfish/issues/40 implemented. Hence some RPC are blocked by that.

**Comment by jingyi2811** [comment]
Working on blockchain.cpp

getchaintips

**Comment by jingyi2811** [comment]
Working on `rpc_tokens.cpp`

getcustomtx
minttokens

**Comment by jingyi2811** [comment]
Working on `rpc_tokens.cpp` while waiting for utxostoaccount to be reviewed

getcustomtx
